packages/next/shared/lib/dynamic.tsx

```diff
@@ -112,6 +112,20 @@ export default function dynamic<P = {}>(
     )
   }
 
+  /**
+   * TODO: Currently, next/dynamic will opt-in to React.lazy if { suspense: true } is used
+   * React 18 will always resolve the Suspense boundary on the server-side, effectively ignoring the ssr option
+   *
+   * In the future, when React Suspense with third-party libraries is stable, we can implement a custom version of
+   * React.lazy that can suspense on the server-side while only loading the component on the client-side
+   */
+  if (loadableOptions.ssr === false && loadableOptions.suspense) {
+    loadableOptions.ssr = true
+    console.warn(
+      `"ssr: false" is ignored by next/dynamic because you can not enable "suspense" while disabling "ssr" at the same time. Read more: https://nextjs.org/docs/messages/invalid-dynamic-suspense`
+    )
```

Comment:
Should we throw an error here instead?

I choose `console.warn` in order to match the current behavior: The existing e2e test case uses `{ ssr: false, suspense: true }` while also `expect(await hasRedbox(browser)).toBe(false)`.

Reply:
I feel the warning is fine since it's not always broken but only when you includes certain browser code. But we can warn them more about it, like other options will also be ignored in suspense mode.
If the checking conditions code are effecting bundle side, we can only error in dev mode so they can get elimated on production.

```
if (process.env.NODE_ENV === 'development') {
   // ...checkOptions
   // ...error if necessary
}
```

Reply:
Done in [`90dc72a` (#39676)](https://github.com/vercel/next.js/pull/39676/commits/90dc72a5a68e904800cdc7a7a3ef545a84f2646b):

- Also add a warning for `loading` option.
- Move test case from e2e to integration
  - "red box" and `console.warn` are dev only